---
title: Gencode exploration
date: 2016-06-04
tags: ["genome"]
---



<div id="gencode-v19" class="section level2">
<h2>Gencode v19</h2>
<p>I downloaded Gencode v19 at <a href="ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_19/gencode.v19.annotation.gtf.gz"><a href="ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_19/gencode.v19.annotation.gtf.gz" class="uri">ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_19/gencode.v19.annotation.gtf.gz</a></a>.</p>
</div>
<div id="genes" class="section level2">
<h2>Genes</h2>
<div id="number" class="section level3">
<h3>Number</h3>
<p>Focusing on autosomes/X/Y, there are 57,783 “genes” of different types:</p>
<p><img src="/post/2016-06-04-GencodeExploration_files/figure-html/genetypes-1.png" width="768" /></p>
<p>I merge the rare types into a <em>other</em> class and some RNAs.</p>
<table>
<thead>
<tr class="header">
<th align="left">gene_type.f</th>
<th align="right">n</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">protein_coding</td>
<td align="right">20332</td>
</tr>
<tr class="even">
<td align="left">pseudogene</td>
<td align="right">13931</td>
</tr>
<tr class="odd">
<td align="left">other</td>
<td align="right">7417</td>
</tr>
<tr class="even">
<td align="left">lincRNA</td>
<td align="right">7114</td>
</tr>
<tr class="odd">
<td align="left">RNA</td>
<td align="right">5934</td>
</tr>
<tr class="even">
<td align="left">miRNA</td>
<td align="right">3055</td>
</tr>
</tbody>
</table>
</div>
<div id="size" class="section level3">
<h3>Size</h3>
<p><img src="/post/2016-06-04-GencodeExploration_files/figure-html/genesize-1.png" width="768" /></p>
<p>The largest annotated genes span more than 2 Mbp:</p>
<table>
<thead>
<tr class="header">
<th align="left">seqnames</th>
<th align="left">gene_name</th>
<th align="left">gene_type</th>
<th align="right">size.Mbp</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">chr7</td>
<td align="left">CNTNAP2</td>
<td align="left">protein_coding</td>
<td align="right">2.304638</td>
</tr>
<tr class="even">
<td align="left">chr9</td>
<td align="left">PTPRD</td>
<td align="left">protein_coding</td>
<td align="right">2.298478</td>
</tr>
<tr class="odd">
<td align="left">chrX</td>
<td align="left">DMD</td>
<td align="left">protein_coding</td>
<td align="right">2.241765</td>
</tr>
<tr class="even">
<td align="left">chr3</td>
<td align="left">LSAMP</td>
<td align="left">protein_coding</td>
<td align="right">2.194861</td>
</tr>
<tr class="odd">
<td align="left">chr11</td>
<td align="left">DLG2</td>
<td align="left">protein_coding</td>
<td align="right">2.172912</td>
</tr>
</tbody>
</table>
<p>The smallest protein-coding annotated genes less than 100 bp:</p>
<table>
<thead>
<tr class="header">
<th align="left">seqnames</th>
<th align="left">gene_name</th>
<th align="left">gene_type</th>
<th align="right">size.bp</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">chr2</td>
<td align="left">AC011308.1</td>
<td align="left">protein_coding</td>
<td align="right">59</td>
</tr>
<tr class="even">
<td align="left">chr12</td>
<td align="left">AC055736.1</td>
<td align="left">protein_coding</td>
<td align="right">66</td>
</tr>
<tr class="odd">
<td align="left">chr16</td>
<td align="left">PIH1</td>
<td align="left">protein_coding</td>
<td align="right">93</td>
</tr>
<tr class="even">
<td align="left">chr2</td>
<td align="left">AC012360.2</td>
<td align="left">protein_coding</td>
<td align="right">96</td>
</tr>
<tr class="odd">
<td align="left">chr5</td>
<td align="left">AC008914.1</td>
<td align="left">protein_coding</td>
<td align="right">102</td>
</tr>
</tbody>
</table>
</div>
<div id="density" class="section level3">
<h3>Density</h3>
<p>Using non-overlapping windows of 1 Mb the gene density looks like this:</p>
<p><img src="/post/2016-06-04-GencodeExploration_files/figure-html/genedens-1.png" width="768" /><img src="/post/2016-06-04-GencodeExploration_files/figure-html/genedens-2.png" width="768" /></p>
<ul>
<li>Chr 19, 17 and 11 have more protein-coding genes than the rest.</li>
<li>Chr Y has more pseudogene compared to other classes.</li>
</ul>
</div>
</div>
<div id="exons" class="section level2">
<h2>Exons</h2>
<div id="number-1" class="section level3">
<h3>Number</h3>
<p>Focusing on autosomes/X/Y, there are 1,196,256 “exons” from different types of genes:</p>
<p><img src="/post/2016-06-04-GencodeExploration_files/figure-html/exontype-1.png" width="768" /></p>
<p>For the rest of the analysis, I use only exons from protein-coding genes.</p>
</div>
<div id="number-per-gene" class="section level3">
<h3>Number per gene</h3>
<p><img src="/post/2016-06-04-GencodeExploration_files/figure-html/exongene-1.png" width="768" /></p>
<table>
<thead>
<tr class="header">
<th align="right">mean.nb.exon</th>
<th align="right">median.nb.exon</th>
<th align="right">max.nb.exon</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">52.9</td>
<td align="right">31</td>
<td align="right">1696</td>
</tr>
</tbody>
</table>
<p>The gene with the most exons is the <a href="http://www.genecards.org/cgi-bin/carddisp.pl?gene=TTN">Titin gene</a>:</p>
<table>
<thead>
<tr class="header">
<th align="left">gene_name</th>
<th align="right">exon</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">TTN</td>
<td align="right">1696</td>
</tr>
<tr class="even">
<td align="left">SYNE1</td>
<td align="right">1377</td>
</tr>
<tr class="odd">
<td align="left">NEB</td>
<td align="right">1225</td>
</tr>
<tr class="even">
<td align="left">CACNA1G</td>
<td align="right">1139</td>
</tr>
<tr class="odd">
<td align="left">CACNA1C</td>
<td align="right">1098</td>
</tr>
</tbody>
</table>
</div>
<div id="size-1" class="section level3">
<h3>Size</h3>
<p>The average exon size is 232 bp.</p>
<p><img src="/post/2016-06-04-GencodeExploration_files/figure-html/exonsize-1.png" width="768" /><img src="/post/2016-06-04-GencodeExploration_files/figure-html/exonsize-2.png" width="768" /></p>
<p>The first exon seems to be slightly larger than the others. I used genes with at least 10 exons to be sure it’s not due to large single-exon genes.</p>
<p>The largest annotated exons are more than 20 Kbp long:</p>
<table>
<thead>
<tr class="header">
<th align="left">seqnames</th>
<th align="left">gene_name</th>
<th align="right">size.kb</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">chr8</td>
<td align="left">TRAPPC9</td>
<td align="right">29.06</td>
</tr>
<tr class="even">
<td align="left">chr5</td>
<td align="left">MCC</td>
<td align="right">27.20</td>
</tr>
<tr class="odd">
<td align="left">chr12</td>
<td align="left">GRIN2B</td>
<td align="right">24.41</td>
</tr>
<tr class="even">
<td align="left">chr19</td>
<td align="left">MUC16</td>
<td align="right">21.69</td>
</tr>
<tr class="odd">
<td align="left">chr2</td>
<td align="left">ABI2</td>
<td align="right">20.55</td>
</tr>
</tbody>
</table>
<p>The smallest are just 1 bp !?</p>
<table>
<thead>
<tr class="header">
<th align="left">seqnames</th>
<th align="left">gene_name</th>
<th align="right">size.bp</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">chr2</td>
<td align="left">ALK</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">chr3</td>
<td align="left">ACAD11</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">chr4</td>
<td align="left">PPA2</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">chr5</td>
<td align="left">PAM</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">chr5</td>
<td align="left">GALNT10</td>
<td align="right">1</td>
</tr>
</tbody>
</table>
</div>
</div>
