---
title: "Stereograms of the 1KGP PCA/tSNE"
date: 2019-11-19
tags: ["R", "plot"]
slug: stereograms-1kgp-pca-tsne
---

```{r include=FALSE}
knitr::opts_chunk$set(message=FALSE, echo=FALSE, warning=FALSE, fig.width=10)
```


```{r libs}
library(ggforce)
library(dplyr)
library(ggplot2)
```

## What is stereoscopy? How to look at a stereogram?

> Technique for creating or enhancing the illusion of depth in an image by means of stereopsis for binocular vision.

--[Stereoscopy (Wikipedia)](https://en.wikipedia.org/wiki/Stereoscopy)

Although some binocular-like devices can be used to merge two images and create the illusion of depth, there are also "freeviewing" techniques that don't require any particular material.
The two main techniques are to either look straight through the image or to cross your eyes.
[Wikihow](https://www.wikihow.com/View-Stereograms) has some tips for the "parallel" method.

I knew about the [autostereogram](https://en.wikipedia.org/wiki/Autostereogram) (I had a book of "magic images" as a kid), like this example from Wikipedia: 

![*Let your eyes see through the image and you'll see a shark appear.*](https://upload.wikimedia.org/wikipedia/commons/8/8f/Stereogram_Tut_Random_Dot_Shark.png)

Side-by-side images were new to me. 
It's the same principle though: either let your eyes look through  or cross your eyes until both images overlap. 
An example from Wikipedia:

![*Cross your eyes to align both images and create a new one in between.*](https://upload.wikimedia.org/wikipedia/commons/thumb/b/bc/Passiflora_caerulea_STEREO_%28R-L%29_2019-06-27.jpg/1280px-Passiflora_caerulea_STEREO_%28R-L%29_2019-06-27.jpg)

For the side-by-side images I find it easier to use the cross-eye technique because it's easy to adjust how much to cross my eyes to overlap both images.

I also noticed that it's important to have my head straight (both eyes are on the same plane).
When using the cross-eyed method, it helps tilting slightly the head on one side or the other to align the two images vertically.

## Making a PCA stereogram using the 1000 Genomes Project data

Now let's see how to make a stereogram of the typical PCA plot using the 1000 Genomes Project data.
We'll use *PC1*/*PC2* for the x/y axis, as usual, and *PC3* for the "depth" dimension.

I downloaded a subset of variants in the 1000 Genomes Project Phase 3 that had at least 10 alternate alleles in called genotypes in autosomes (0.1% of variants with *AC>=10*). 
I used the autosomes' VCFs at `ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.<CHR>.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz`.
Homozygous refs were then converted to *0*, heterozygous variants to *1* and homozygous alts to *2*.

```{r pca1kgp}
pop.info = tibble(population=c('CHB', 'JPT', 'CHS', 'CDX', 'KHV', 'CEU', 'TSI', 'FIN', 'GBR', 'IBS', 'YRI', 'LWK', 'GWD', 'MSL', 'ESN', 'ASW', 'ACB', 'MXL', 'PUR', 'CLM', 'PEL', 'GIH', 'PJL', 'BEB', 'STU', 'ITU'),
                  super.population=c('EAS', 'EAS', 'EAS', 'EAS', 'EAS', 'EUR', 'EUR', 'EUR', 'EUR', 'EUR', 'AFR', 'AFR', 'AFR', 'AFR', 'AFR', 'AFR', 'AFR', 'AMR', 'AMR', 'AMR', 'AMR', 'SAS', 'SAS', 'SAS', 'SAS', 'SAS'))

info = read.csv('smalldata/20130606_sample_info.csv', as.is=TRUE) %>%
  mutate(sample=Sample, population=Population, pop.desc=Population.Description) %>%
  select(sample, population, pop.desc) %>% merge(pop.info)

if(!file.exists('smalldata/1kgp-pca.RData')){
  ##   curl ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr22.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz | zcat | head -1000 | grep '#' --color=never > ss.vcf
  ## for CHR in `seq 1 22`
  ## do
  ##     curl ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr$CHR.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz | zcat | grep -v '##' |  awk 'BEGIN{srand()}{match($0, /AC=([^;]*);/, ac); if(ac[1]>9){if(rand()<.001){print $0}}}' >> ss.vcf
  ## done
  library(VariantAnnotation)
  vcf = readVcf('ss.vcf')
  gt = geno(vcf)$GT
  convGT = c('0|0'=0, '.|.'=0, '1|0'=1, '0|1'=1, '1|1'=2)
  gt = apply(gt, 2, function(x)convGT[x])
  gt = gt[which(rowSums(gt)>0),]
  pca.o = prcomp(t(gt))
  pca.o$x = pca.o$x[,1:100]
  pca.o$sdev = pca.o$sdev[1:100]
  pca.o$rotation = pca.o$rotation[1:100,]
  save(pca.o, file='smalldata/1kgp-pca.RData')
} else {
  load('smalldata/1kgp-pca.RData')
}
pca.df = tibble(sample=rownames(pca.o$x)) %>%
  cbind(as.data.frame(pca.o$x[,1:10])) %>% merge(info)
```

The typical PCA looks like this:

```{r normalpcas, echo=TRUE}
ggplot(pca.df, aes(x=PC1, y=PC2, color=super.population)) +
  geom_point(alpha=.4) + theme_bw() +
  scale_color_brewer(palette='Dark2') + 
  guides(colour=guide_legend(override.aes=list(alpha=1)))
ggplot(pca.df, aes(x=PC1, y=PC3, color=super.population)) +
  geom_point(alpha=.4) + theme_bw() +
  scale_color_brewer(palette='Dark2') + 
  guides(colour=guide_legend(override.aes=list(alpha=1)))
```

Now the same data with a stereogram using the [ggforce R package](https://ggforce.data-imaginist.com/):

```{r stereopca, echo=TRUE, fig.height=6}
## add some white space around the points
extra.margin = .2
pc1.r = range(pca.df$PC1)
pc1.r = pc1.r + c(-1,1)*extra.margin*diff(pc1.r)
pc2.r = range(pca.df$PC2)
pc2.r = pc2.r + c(-1,1)*extra.margin*diff(pc2.r)
##
ggplot(pca.df, aes(x=PC1, y=PC2, depth=-PC3, color=super.population)) +
  geom_point(alpha=.4) + theme_bw() +
  scale_color_brewer(palette='Dark2') +
  guides(colour=guide_legend(override.aes=list(alpha=1))) + 
  facet_stereo(IPD=-63.5) + theme(legend.position='bottom') +
  xlim(pc1.r) + ylim(pc2.r)
```

A few notes: 

- I specified the limits of the graph to add a bit of white space around the points. It helps a bit IMO.
- I specified `IPD=-63.5` to switch to "cross-eyed' mode. The default is `IPD=63.5` for the "parallel" technique. I could only manage the "parallel" viewing by making the plot very small.

The result is pretty cool, we can tell that the *SAS* are in front of all other points, and that the branch of *AMR* is stretching to the back.

It's fun to imagine showing these graphs during a presentation and watching a crossed-eyed audience like:

![](../../../../imgs/crosseyeddenis.gif)

## Making a tSNE stereogram using the 1000 Genomes Project data

A tSNE with 3 components using the top 30 PCs looks like this:

```{r tsne}
library(Rtsne)

set.seed(12345)
tsne.o = Rtsne(pca.o$x[,1:30], pca=FALSE, dims=3)

tsne.df = tibble(sample=rownames(pca.o$x),
                 tsne1=tsne.o$Y[,1],
                 tsne2=tsne.o$Y[,2],
                 tsne3=tsne.o$Y[,3]) %>% merge(info)

## add some white space around the points
extra.margin = .2
tsne1.r = range(tsne.df$tsne1)
tsne1.r = tsne1.r + c(-1,1)*extra.margin*diff(tsne1.r)
tsne2.r = range(tsne.df$tsne2)
tsne2.r = tsne2.r + c(-1,1)*extra.margin*diff(tsne2.r)
##
ggplot(tsne.df, aes(tsne1, tsne2, depth=tsne3, color=population)) +
  geom_point(alpha=.4, size=2) + facet_stereo(IPD=-63.5) +
  theme_bw() + theme(legend.position='bottom') +
  xlim(tsne1.r) + ylim(tsne2.r) +
  guides(color=guide_legend(override.aes=list(alpha=1), ncol=15))
```

The blobs are nicely separated in 3D.
We can clearly see some blobs corresponding to single populations.
